File: leetcode/0040.Combination-Sum-II/CombinationSumII.py
from typing import List

class Solution:

    # Runtime: 40 ms, faster than 97.22% of Python3 online submissions for Combination Sum II.
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        def combinHelper(candidates, target, prefixRes, res):
            for i, n in enumerate(candidates):
                if i > 0 and n == candidates[i-1]:
                    continue
                if n == target:
                    res.append(prefixRes + [n])
                elif n < target:
                    combinHelper(candidates[i+1:], target - n, prefixRes + [n], res)
                else:
                    break
        res = []
        combinHelper(candidates, target, [], res)
        return res

File: leetcode/0040.Combination-Sum-II/CombinationSumII.go
package leetcode

import "sort"

func combinHelper(candidates []int, target int, prefixRes []int, res *[][]int) {
	for i, n := range candidates {
		if i > 0 && n == candidates[i-1] {
			continue
		}

		if n == target {
			copyPrefix := make([]int, len(prefixRes))
			copy(copyPrefix, prefixRes)
			*res = append(*res, append(copyPrefix, n))
		} else if n < target {
			combinHelper(candidates[i+1:], target-n, append(prefixRes, n), res)
		} else {
			break
		}
	}
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Combination Sum II.
func combinationSum2(candidates []int, target int) [][]int {
	res := [][]int{}
	sort.Ints(candidates)
	combinHelper(candidates, target, []int{}, &res)
	return res
}
